File: vue3/src/stores/conversation.ts
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import { conversationApi, type Conversation, type Message } from '@/api/conversation'

export const useConversationStore = defineStore('conversation', () => {
  // 状态
  const conversations = ref<Conversation[]>([])
  const currentConversation = ref<Conversation | null>(null)
  const currentMessages = ref<Message[]>([])
  const loading = ref(false)

  // 计算属性
  const hasConversations = computed(() => conversations.value.length > 0)

  // 获取对话列表
  const loadConversations = async (limit: number = 50, offset: number = 0) => {
    loading.value = true
    try {
      const response = await conversationApi.getConversations(limit, offset)
      // 处理可能的 null 值
      conversations.value = response.conversations.map(conv => ({
        id: conv.id,
        user_id: conv.user_id,
        title: conv.title || '',
        created_at: conv.created_at || '',
        updated_at: conv.updated_at || '',
        message_count: conv.message_count
      }))
      return response
    } catch (error) {
      console.error('加载对话列表失败:', error)
      throw error
    } finally {
      loading.value = false
    }
  }

  // 获取对话详情（包含消息）
  const loadConversation = async (conversationId: number) => {
    loading.value = true
    try {
      const detail = await conversationApi.getConversation(conversationId)
      currentConversation.value = {
        id: detail.id,
        user_id: detail.user_id,
        title: detail.title || '',
        created_at: detail.created_at || '',
        updated_at: detail.updated_at || '',
        message_count: detail.message_count
      }
      // 确保消息列表不为空，并处理可能的 null 值
      currentMessages.value = (detail.messages || []).map(msg => ({
        id: msg.id,
        role: msg.role,
        content: msg.content,
        created_at: msg.created_at || new Date().toISOString()
      }))
      return detail
    } catch (error) {
      console.error('加载对话详情失败:', error)
      throw error
    } finally {
      loading.value = false
    }
  }

  // 创建新对话
  const createConversation = async (title?: string): Promise<Conversation> => {
    try {
      const conversation = await conversationApi.createConversation(title)
      // 处理可能的 null 值
      const normalizedConv: Conversation = {
        id: conversation.id,
        user_id: conversation.user_id,
        title: conversation.title || '',
        created_at: conversation.created_at || new Date().toISOString(),
        updated_at: conversation.updated_at || new Date().toISOString(),
        message_count: conversation.message_count
      }
      // 添加到列表开头
      conversations.value.unshift(normalizedConv)
      return normalizedConv
    } catch (error) {
      console.error('创建对话失败:', error)
      throw error
    }
  }

  // 设置当前对话
  const setCurrentConversation = (conversation: Conversation | null) => {
    currentConversation.value = conversation
    if (conversation) {
      loadConversation(conversation.id)
    } else {
      currentMessages.value = []
    }
  }

  // 添加消息到当前对话
  const addMessage = (message: Message) => {
    currentMessages.value.push(message)
  }

  // 更新对话标题
  const updateConversationTitle = async (conversationId: number, title: string) => {
    try {
      await conversationApi.updateConversationTitle(conversationId, title)
      // 更新本地状态
      const conversation = conversations.value.find(c => c.id === conversationId)
      if (conversation) {
        conversation.title = title
      }
      if (currentConversation.value?.id === conversationId) {
        currentConversation.value.title = title
      }
    } catch (error) {
      console.error('更新对话标题失败:', error)
      throw error
    }
  }

  // 删除对话
  const deleteConversation = async (conversationId: number) => {
    try {
      await conversationApi.deleteConversation(conversationId)
      // 从列表中移除
      conversations.value = conversations.value.filter(c => c.id !== conversationId)
      // 如果删除的是当前对话，清空当前对话
      if (currentConversation.value?.id === conversationId) {
        currentConversation.value = null
        currentMessages.value = []
      }
    } catch (error) {
      console.error('删除对话失败:', error)
      throw error
    }
  }

  // 刷新对话列表
  const refreshConversations = async () => {
    await loadConversations()
  }

  // 清空状态
  const clear = () => {
    conversations.value = []
    currentConversation.value = null
    currentMessages.value = []
  }

  return {
    // 状态
    conversations,
    currentConversation,
    currentMessages,
    loading,
    // 计算属性
    hasConversations,
    // 方法
    loadConversations,
    loadConversation,
    createConversation,
    setCurrentConversation,
    addMessage,
    updateConversationTitle,
    deleteConversation,
    refreshConversations,
    clear
  }
})

File: python/app/routers/conversation.py
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel
from typing import Optional, List
from app.utils.response import ResponseWrapper as R
from app.utils.status_code import StatusCode
from app.db.conversation_dao import (
    create_conversation,
    get_conversations_by_user_id,
    get_conversation_by_id,
    update_conversation_title,
    delete_conversation,
    get_conversation_count_by_user_id
)
from app.db.message_dao import get_messages_by_conversation_id, get_message_count_by_conversation_id
from app.dependencies.auth import get_current_user
from app.db.models.user import User
from app.db.models.conversation import Conversation

router = APIRouter()


class CreateConversationRequest(BaseModel):
    title: Optional[str] = ""


class UpdateConversationTitleRequest(BaseModel):
    title: str


class ConversationResponse(BaseModel):
    id: int
    user_id: int
    title: str
    created_at: str
    updated_at: str
    message_count: Optional[int] = None


@router.get("/conversations")
def get_conversations(
    limit: int = 50,
    offset: int = 0,
    current_user: User = Depends(get_current_user)
):
    """
    获取当前用户的对话列表
    
    Args:
        limit: 返回数量限制（默认50）
        offset: 偏移量（默认0）
        current_user: 当前用户（通过认证依赖注入）
        
    Returns:
        成功响应，包含对话列表
    """
    try:
        conversations = get_conversations_by_user_id(current_user.id, limit=limit, offset=offset)
        total = get_conversation_count_by_user_id(current_user.id)
        
        conversation_list = []
        for conv in conversations:
            # 计算每个对话的消息数量
            message_count = get_message_count_by_conversation_id(conv.id)
            conversation_list.append({
                "id": conv.id,
                "user_id": conv.user_id,
                "title": conv.title,
                "created_at": conv.created_at.isoformat() if conv.created_at else None,
                "updated_at": conv.updated_at.isoformat() if conv.updated_at else None,
                "message_count": message_count
            })
        
        return R.success(data={
            "conversations": conversation_list,
            "total": total,
            "limit": limit,
            "offset": offset
        }, msg="获取对话列表成功")
        
    except Exception as e:
        return R.error(msg=f"获取对话列表失败: {str(e)}", code=StatusCode.FAIL)


@router.post("/conversations")
def create_conversation_endpoint(
    request: CreateConversationRequest,
    current_user: User = Depends(get_current_user)
):
    """
    创建新对话
    
    Args:
        request: 创建对话请求（title可选）
        current_user: 当前用户（通过认证依赖注入）
        
    Returns:
        成功响应，包含对话信息
    """
    try:
        conversation = create_conversation(
            user_id=current_user.id,
            title=request.title or ""
        )
        
        return R.success(data={
            "id": conversation.id,
            "user_id": conversation.user_id,
            "title": conversation.title,
            "created_at": conversation.created_at.isoformat() if conversation.created_at else None,
            "updated_at": conversation.updated_at.isoformat() if conversation.updated_at else None
        }, msg="创建对话成功")
        
    except Exception as e:
        return R.error(msg=f"创建对话失败: {str(e)}", code=StatusCode.FAIL)


@router.get("/conversations/{conversation_id}")
def get_conversation(
    conversation_id: int,
    current_user: User = Depends(get_current_user)
):
    """
    获取对话详情和消息列表
    
    Args:
        conversation_id: 对话ID
        current_user: 当前用户（通过认证依赖注入）
        
    Returns:
        成功响应，包含对话信息和消息列表
    """
    try:
        conversation = get_conversation_by_id(conversation_id)
        
        if not conversation:
            return R.error(msg="对话不存在", code=StatusCode.CONVERSATION_NOT_FOUND)
        
        # 验证所有权
        if conversation.user_id != current_user.id:
            return R.error(msg="无权访问该对话", code=StatusCode.CONVERSATION_ACCESS_DENIED)
        
        # 获取消息列表
        messages = get_messages_by_conversation_id(conversation_id)
        message_list = []
        for msg in messages:
            message_list.append({
                "id": msg.id,
                "role": msg.role,
                "content": msg.content,
                "created_at": msg.created_at.isoformat() if msg.created_at else None
            })
        
        return R.success(data={
            "id": conversation.id,
            "user_id": conversation.user_id,
            "title": conversation.title,
            "created_at": conversation.created_at.isoformat() if conversation.created_at else None,
            "updated_at": conversation.updated_at.isoformat() if conversation.updated_at else None,
            "messages": message_list
        }, msg="获取对话详情成功")
        
    except Exception as e:
        return R.error(msg=f"获取对话详情失败: {str(e)}", code=StatusCode.FAIL)


@router.put("/conversations/{conversation_id}/title")
def update_conversation_title_endpoint(
    conversation_id: int,
    request: UpdateConversationTitleRequest,
    current_user: User = Depends(get_current_user)
):
    """
    更新对话标题
    
    Args:
        conversation_id: 对话ID
        request: 更新标题请求
        current_user: 当前用户（通过认证依赖注入）
        
    Returns:
        成功响应
    """
    try:
        conversation = get_conversation_by_id(conversation_id)
        
        if not conversation:
            return R.error(msg="对话不存在", code=StatusCode.CONVERSATION_NOT_FOUND)
        
        # 验证所有权
        if conversation.user_id != current_user.id:
            return R.error(msg="无权访问该对话", code=StatusCode.CONVERSATION_ACCESS_DENIED)
        
        success = update_conversation_title(conversation_id, request.title)
        
        if not success:
            return R.error(msg="更新对话标题失败", code=StatusCode.FAIL)
        
        return R.success(data=None, msg="更新对话标题成功")
        
    except Exception as e:
        return R.error(msg=f"更新对话标题失败: {str(e)}", code=StatusCode.FAIL)


@router.delete("/conversations/{conversation_id}")
def delete_conversation_endpoint(
    conversation_id: int,
    current_user: User = Depends(get_current_user)
):
    """
    删除对话（级联删除所有消息）
    
    Args:
        conversation_id: 对话ID
        current_user: 当前用户（通过认证依赖注入）
        
    Returns:
        成功响应
    """
    try:
        conversation = get_conversation_by_id(conversation_id)
        
        if not conversation:
            return R.error(msg="对话不存在", code=StatusCode.CONVERSATION_NOT_FOUND)
        
        # 验证所有权
        if conversation.user_id != current_user.id:
            return R.error(msg="无权访问该对话", code=StatusCode.CONVERSATION_ACCESS_DENIED)
        
        success = delete_conversation(conversation_id)
        
        if not success:
            return R.error(msg="删除对话失败", code=StatusCode.FAIL)
        
        return R.success(data=None, msg="删除对话成功")
        
    except Exception as e:
        return R.error(msg=f"删除对话失败: {str(e)}", code=StatusCode.FAIL)

